src/grift/insert-casts.rkt

```diff
@@ -71,30 +52,96 @@
             (match-define (cons loc exp) loc.exp)
             (printf "\t\t~a\n" (srcloc->string loc))))))
     (error 'Type-Error err-str))
-  
-  (debug (Prog (list name (unique-counter-next! uc) type) new-tl*)))
 
-(: mk-cast ((-> Blame-Label) C0-Expr Grift-Type Grift-Type -> C0-Expr))
+  (debug who (Prog (list name (unique-counter-next! uc) type) new-tl*)))
+
+
+;; A list of casts that are inserted and their source location
+(: casts-inserted (Boxof (Listof (cons srcloc C0-Expr))))
+(define casts-inserted (box '()))
+(: add-cast! : srcloc C0-Expr -> Void)
+(define (add-cast! s e)
+  (set-box! casts-inserted (cons (cons s e) (unbox casts-inserted))))
+(: no-casts! : -> Void)
+(define (no-casts!)
+  (set-box! casts-inserted '()))
+
+
+;; is `t1` a subtype of `t2`?
+(define (subtype-of? t1 t2)
```

Comment:
I agree that this function is misnamed with what it is currently doing, but to my understanding we could and probably should perform a subtyping check here. Such a subtyping function would have to worry about equirecursive equality. It might be better to just do the conversion now, and make this function actually perform `subtype-of?`. I think the conversion amounts to making sure the variance is correct and making the tuple check "arity less than".

Reply:
I ended up making this compute a proper subtype.